Implement std::error::Error on cbindgen::Error to support Failure

Currently `cbindgen::Error` can't be used with failure because `std::error::Error` isn't implemented on `cbindgen::Error`. 

Example code (using `ResultExt::context(..., ...)` instead of `.context(..) to make rustc print a clearer error message):

```
use cbindgen;
use failure::ResultExt;

fn test() {
    let result = cbindgen::Builder::new().with_crate(".").generate();
    ResultExt::context(result, "Failed to run cbindgen");
}
```

```
error[E0277]: the trait bound `cbindgen::bindgen::error::Error: std::error::Error` is not satisfied
 --> src/error.rs:6:5
  |
6 |     ResultExt::context(result, "Failed to run cbindgen");
  |     ^^^^^^^^^^^^^^^^^^ the trait `std::error::Error` is not implemented for `cbindgen::bindgen::error::Error`
  |
  = note: required because of the requirements on the impl of `failure::Fail` for `cbindgen::bindgen::error::Error`
  = note: required because of the requirements on the impl of `failure::result_ext::ResultExt<cbindgen::bindgen::bindings::Bindings, cbindgen::bindgen::error::Error>` for `std::result::Result<cbindgen::bindgen::bindings::Bindings, cbindgen::bindgen::error::Error>`
  = note: required by `failure::result_ext::ResultExt::context`
```

--- a/src/bindgen/cargo/cargo_expand.rs
+++ b/src/bindgen/cargo/cargo_expand.rs
@@ -3,6 +3,8 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::env;
+use std::error;
+use std::fmt;
 use std::io;
 use std::path::Path;
 use std::process::Command;
@@ -30,6 +32,26 @@ impl From<io::Error> for Error {
 impl From<Utf8Error> for Error {
     fn from(err: Utf8Error) -> Self {
         Error::Utf8(err)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Io(ref err) => err.fmt(f),
+            Error::Utf8(ref err) => err.fmt(f),
+            Error::Compile(ref err) => write!(f, "{}", err),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            Error::Io(ref err) => Some(err),
+            Error::Utf8(ref err) => Some(err),
+            Error::Compile(..) => None,
+        }
     }
 }
 

--- a/src/bindgen/cargo/cargo_metadata.rs
+++ b/src/bindgen/cargo/cargo_metadata.rs
@@ -11,6 +11,8 @@
 
 use std::collections::HashMap;
 use std::env;
+use std::error;
+use std::fmt;
 use std::io;
 use std::path::Path;
 use std::process::Command;
@@ -100,6 +102,26 @@ impl From<Utf8Error> for Error {
 impl From<serde_json::Error> for Error {
     fn from(err: serde_json::Error) -> Self {
         Error::Json(err)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Io(ref err) => err.fmt(f),
+            Error::Utf8(ref err) => err.fmt(f),
+            Error::Json(ref err) => err.fmt(f),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            Error::Io(ref err) => Some(err),
+            Error::Utf8(ref err) => Some(err),
+            Error::Json(ref err) => Some(err),
+        }
     }
 }
 

--- a/src/bindgen/cargo/cargo_toml.rs
+++ b/src/bindgen/cargo/cargo_toml.rs
@@ -2,6 +2,8 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+use std::error;
+use std::fmt;
 use std::fs::File;
 use std::io;
 use std::io::Read;
@@ -26,6 +28,24 @@ impl From<io::Error> for Error {
 impl From<toml::de::Error> for Error {
     fn from(err: toml::de::Error) -> Self {
         Error::Toml(err)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Error::Io(ref err) => err.fmt(f),
+            Error::Toml(ref err) => err.fmt(f),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            Error::Io(ref err) => Some(err),
+            Error::Toml(ref err) => Some(err),
+        }
     }
 }
 

--- a/src/bindgen/error.rs
+++ b/src/bindgen/error.rs
@@ -2,6 +2,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+use std::error;
 use std::fmt;
 
 pub use bindgen::cargo::cargo_expand::Error as CargoExpandError;
@@ -69,6 +70,18 @@ impl fmt::Display for Error {
                 "Parsing crate `{}`: cannot open file `{}`.",
                 crate_name, src_path
             ),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            Error::CargoMetadata(_, ref error) => Some(error),
+            Error::CargoToml(_, ref error) => Some(error),
+            Error::CargoExpand(_, ref error) => Some(error),
+            Error::ParseSyntaxError { ref error, .. } => Some(error),
+            Error::ParseCannotOpenFile { .. } => None,
         }
     }
 }

